Sharded config

jordanwheeler opened the pull request:
Not sure how to approach this, so I'm just gonna open a PR in my fork of LHM for now.

This gets the slave lag throttler working with Shopify.  The previous version of this only checked the master's slave, but this will walk the replication chain and check the slave lag for every node in a shard.  I'm not sure if we can get generic enough to merge this to the main soundcloud repo, but I've tested it extensively in Shopify and it works for us for both sharded and master only LHMs.  

At a certain point in the chain, a node ends up replicating to hadoop, etc., and when this tries to connect  a MySql2 error is raised because the credentials we're using don't work there.  We decided the best approach here is to just connect to everything in the chain, and if an error is raised on connection, we simply log it, skip that node, and move on.

The config we get from ActiveRecord will connect to the master shard by default.  To get the config for other shards, I'm getting the shard name from [shopify core](https://github.com/Shopify/shopify/blob/master/lib/sharding/switcher.rb#L32-L34).  In the LHM migration, I pass our `current_shard_name` method (it returns `'shard_1'`, etc), and that'll let us determine what config we should be using.

Comments please

Commit:
Detect slave lag across entire replication chain

camilo commented on lib/lhm/throttler/slave_lag.rb:
```diff
@@ -15,7 +15,7 @@ def initialize(options = {})
         @stride = options[:stride] || DEFAULT_STRIDE
         @allowed_lag = options[:allowed_lag] || DEFAULT_MAX_ALLOWED_LAG
-        @slave_connections = {}
+        @get_current_shard = options[:get_current_shard]
```
why not just call this get config or something like that, the shard part is orthogonal?

shuhaowu commented on lib/lhm/throttler/slave_lag.rb:
```diff
@@ -44,14 +44,30 @@ def throttle_seconds
+      end
+
+      def get_slave_hosts
+        slave_hosts = []
```
nitpick, the fact that this variable is the same name as one of the methods here might be a little confusing if you missed this declaration line in the first pass.

shuhaowu commented on lib/lhm/throttler/slave_lag.rb:
```diff
@@ -75,9 +91,19 @@ def slave_lag(slave)
+
+      def slave_config(slave)
         config = ActiveRecord::Base.connection_pool.spec.config.dup
+        config = config[@get_current_shard.call.to_sym] if @get_current_shard && @get_current_shard.call
```
Is this assuming that the `config` hash has nested configurations of each shard? This seems like it's pretty specific.

shuhaowu commented:
Tell me if I'm wrong because I feel like I'm misunderstanding something: 

So the idea here is that when getting a list of slave hosts, we first look at at the master and get all the slaves, and then from those slaves we get the slaves under them in a breath first search. In order to get the slave connections, a lambda has been passed into the throttler which makes it get the configuration of a shard assuming that that configuration is nested under the default `ActiveRecord::Base.connection_pool.spec.config`?

Why not just do `get_current_config` instead? It would be much more generic. You could also pass that method a slave host which would be even more generic (if slaves have different password for some, let's say).

jordanwheeler commented:
@shuhaowu @camilo @jasonhl I've refactored this and added a `Slave` class.  Let me know what you think.  It definitely works across multiple shards and the code is nicer IMHO.  

It's much easier working with `Mysql2::Client` to connect to each slave since you don't need to mess with the existing ActiveRecord connections and connection pools.  Which also means we don't need to change much in our codebase.
